src/puppetlabs/trapperkeeper/internal.clj

```diff
@@ -448,20 +475,29 @@
         this)
       (a/stop [this]
         (shutdown! app-context)
-        this))))
+        this)
+      (a/restart [this]
+        (try
+          (run-lifecycle-fns app-context s/stop "stop" (reverse ordered-services))
```

Comment:
@camlow325 I think the normal TK shutdown logic is why we coded it this way... relying on TK to call the remaining `stop` fns.  As for double-stop, I'd definitely prefer to *not* impose a requirement that services be resilient to stop being called twice, but I'm not sure I see an easy way to achieve that, so, not sure it's worth the ROI?  Happy to discuss that further, though...

We kind of have a similar issue already, in that if an error occurs during startup, then the `stop` functions may be called w/o `init` or `start` being executed successfully, meaning that services already (unfortunately) have some responsibility to implement their `stop` fns to be resilient to that... I think we have a ticket somewhere about some services being subject to an NPE in this situation today.  So I guess it doesn't feel like double-stop adds much more burden than we already have due to that issue...